File: db2-sidecar/ConnectionManager.cs
// db2-sidecar/ConnectionManager.cs
using IBM.Data.DB2.Core;

namespace Db2Sidecar;

public class ConnectionManager
{
    private readonly Dictionary<string, DB2Connection> _connections = new();

    /// <summary>
    /// Opens a DB2 connection and stores it under connId.
    /// Returns null on success, or an error message on failure.
    /// </summary>
    public async Task<string?> Connect(string connId, string connString, string username, string password)
    {
        try
        {
            // IBM.Data.Db2.Core accepts UID/PWD in the connection string
            var fullConnStr = $"{connString};UID={username};PWD={password};";
            var conn = new DB2Connection(fullConnStr);
            await conn.OpenAsync();
            _connections[connId] = conn;
            return null;
        }
        catch (Exception ex)
        {
            return ex.Message;
        }
    }

    /// <summary>Returns the live connection, or null if not found.</summary>
    public DB2Connection? Get(string connId) =>
        _connections.TryGetValue(connId, out var conn) ? conn : null;

    /// <summary>Closes and removes a connection. No-op if not found.</summary>
    public void Disconnect(string connId)
    {
        if (_connections.Remove(connId, out var conn))
        {
            conn.Close();
            conn.Dispose();
        }
    }
}

File: db2-sidecar-java/src/main/java/com/aiterm/db2sidecar/ConnectionManager.java
package com.aiterm.db2sidecar;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.HashMap;
import java.util.Map;

public class ConnectionManager {
    private final Map<String, Connection> connections = new HashMap<>();

    /**
     * Opens a JDBC connection and stores it under connId.
     * connString is a JDBC URL: jdbc:db2://host:port/database
     * Returns null on success, or an error message on failure.
     */
    public String connect(String connId, String connString, String username, String password) {
        try {
            Class.forName("com.ibm.db2.jcc.DB2Driver");
            Connection conn = DriverManager.getConnection(connString, username, password);
            Connection existing = connections.put(connId, conn);
            if (existing != null) {
                try { existing.close(); } catch (SQLException ignored) {}
            }
            return null;
        } catch (ClassNotFoundException e) {
            return "IBM JDBC driver not found: " + e.getMessage();
        } catch (SQLException e) {
            return formatSqlError(e);
        }
    }

    /** Returns the live connection, or null if not found. */
    public Connection get(String connId) {
        return connections.get(connId);
    }

    /** Closes and removes a connection. No-op if not found. */
    public void disconnect(String connId) {
        Connection conn = connections.remove(connId);
        if (conn != null) {
            try { conn.close(); } catch (SQLException ignored) {}
        }
    }

    private String formatSqlError(SQLException e) {
        StringBuilder sb = new StringBuilder(e.getMessage());
        sb.append(" [SQLSTATE=").append(e.getSQLState())
          .append(" ErrorCode=").append(e.getErrorCode()).append("]");
        return sb.toString();
    }
}
